From cd6557d21815618a2ba6d27d8a2015575a854fd1 Mon Sep 17 00:00:00 2001
From: Phil Sutter <psutter@redhat.com>
Date: Tue, 28 Feb 2017 16:12:21 +0100
Subject: [PATCH] bridge: add support for dynamic fdb entries

Bugzilla: https://bugzilla.redhat.com/show_bug.cgi?id=1417289
Upstream Status: iproute2.git commit c6d0cfb54b8b4

commit c6d0cfb54b8b467a57d259f5c9cec91f94f4e59e
Author: Roopa Prabhu <roopa@cumulusnetworks.com>
Date:   Fri Feb 19 21:34:52 2016 -0800

    bridge: add support for dynamic fdb entries

    This patch is a follow up to the recently added
    'static' fdb option.

    It introduces a new option 'dynamic' which adds
    dynamic fdb entries with NUD_REACHABLE.

    $bridge fdb add 00:01:02:03:04:06 dev eth0 master dynamic

    $bridge fdb show
    00:01:02:03:04:06 dev eth0

    This patch also documents all fdb types. Removes 'temp'
    from usage message since it is now replaced by 'static'.
    'temp' still works and is synonymous with static.

    Signed-off-by: Wilson Kok <wkok@cumulusnetworks.com>
    Signed-off-by: Roopa Prabhu <roopa@cumulusnetworks.com>
---
 bridge/fdb.c      |  5 ++++-
 man/man8/bridge.8 | 14 +++++++++++++-
 2 files changed, 17 insertions(+), 2 deletions(-)

diff --git a/bridge/fdb.c b/bridge/fdb.c
index cb07345..0222fd6 100644
--- a/bridge/fdb.c
+++ b/bridge/fdb.c
@@ -32,7 +32,7 @@ static void usage(void)
 {
 	fprintf(stderr, "Usage: bridge fdb { add | append | del | replace } ADDR dev DEV\n"
 			"              [ self ] [ master ] [ use ] [ router ]\n"
-			"              [ local | temp | static ] [ dst IPADDR ] [ vlan VID ]\n"
+			"              [ local | static | dynamic ] [ dst IPADDR ] [ vlan VID ]\n"
 		        "              [ port PORT] [ vni VNI ] [ via DEV ]\n");
 	fprintf(stderr, "       bridge fdb [ show [ br BRDEV ] [ brport DEV ] ]\n");
 	exit(-1);
@@ -303,6 +303,9 @@ static int fdb_modify(int cmd, int flags, int argc, char **argv)
 		} else if (matches(*argv, "temp") == 0 ||
 			   matches(*argv, "static") == 0) {
 			req.ndm.ndm_state |= NUD_REACHABLE;
+		} else if (matches(*argv, "dynamic") == 0) {
+			req.ndm.ndm_state |= NUD_REACHABLE;
+			req.ndm.ndm_state &= ~NUD_NOARP;
 		} else if (matches(*argv, "vlan") == 0) {
 			if (vid >= 0)
 				duparg2("vlan", *argv);
diff --git a/man/man8/bridge.8 b/man/man8/bridge.8
index 864f983..0a36e75 100644
--- a/man/man8/bridge.8
+++ b/man/man8/bridge.8
@@ -54,7 +54,7 @@ bridge \- show / manipulate bridge addresses and devices
 .I LLADDR
 .B dev
 .IR DEV " { "
-.BR local " | " temp " } [ "
+.BR local " | " static " | " dynamic " } [ "
 .BR self " ] [ " master " ] [ " router " ] [ " use " ] [ "
 .B dst
 .IR IPADDR " ] [ "
@@ -338,6 +338,18 @@ the Ethernet MAC address.
 .BI dev " DEV"
 the interface to which this address is associated.
 
+.B local
+- is a local permanent fdb entry
+.sp
+
+.B static
+- is a static (no arp) fdb entry
+.sp
+
+.B dynamic
+- is a dynamic reachable age-able fdb entry
+.sp
+
 .B self
 - the address is associated with the port drivers fdb. Usually hardware.
 .sp
-- 
1.8.3.1

